syft/pkg/cataloger/binary/default_classifiers.go

```diff
@@ -120,6 +120,24 @@ var defaultClassifiers = []classifier{
 		Package: "busybox",
 		CPEs:    singleCPE("cpe:2.3:a:busybox:busybox:*:*:*:*:*:*:*:*"),
 	},
+	{
+		Class:    "haproxy-binary-variant1",
+		FileGlob: "**/haproxy",
+		EvidenceMatcher: fileContentsVersionMatcher(
+			`(?m)HA-Proxy version (?P<version>[0-9]+\.[0-9]+\.[0-9]+)`),
+		Package: "haproxy",
+		PURL:    mustPURL("pkg:generic/haproxy@version"),
+		CPEs:    singleCPE("cpe:2.3:a:haproxy:haproxy:*:*:*:*:*:*:*:*"),
+	},
+	{
+		Class:    "haproxy-binary-variant2",
+		FileGlob: "**/haproxy",
+		EvidenceMatcher: fileContentsVersionMatcher(
+			`(?m)(?P<version>[0-9]+\.[0-9]+\.[0-9]+)-[0-9a-zA-Z]{7}.+HAProxy version`),
+		Package: "haproxy",
+		PURL:    mustPURL("pkg:generic/haproxy@version"),
+		CPEs:    singleCPE("cpe:2.3:a:haproxy:haproxy:*:*:*:*:*:*:*:*"),
+	},
```

Comment:
It might be nice to combine these into a single classifier, maybe something like (all of this is completely untested):
```go
	{
		Class:    "haproxy-binary",
		FileGlob: "**/haproxy",
		EvidenceMatcher: fileContentsVersionMatchers(
			`(?m)HA-Proxy version (?P<version>[0-9]+\.[0-9]+\.[0-9]+)`
			`(?m)(?P<version>[0-9]+\.[0-9]+\.[0-9]+)-[0-9a-zA-Z]{7}.+HAProxy version`),
		Package: "haproxy",
		PURL:    mustPURL("pkg:generic/haproxy@version"),
		CPEs:    singleCPE("cpe:2.3:a:haproxy:haproxy:*:*:*:*:*:*:*:*"),
	},
```

If we added a new classifier function, something like this:
```go
func fileContentsVersionMatchers(patterns ...string) evidenceMatcher {
	var matchers []evidenceMatcher
	for _, pattern := range patterns {
		matchers = append(matchers, fileContentsVersionMatcher(pattern))
	}
	return func(classifier classifier, reader source.LocationReadCloser) ([]pkg.Package, error) {
		for _, matcher := range matchers {
			pkgs, err := matcher(classifier, reader)
			if err != nil {
				return nil, err
			}
			if pkgs != nil {
				return pkgs, nil
			}
		}
		return nil, nil
	}
}
```

Another option is just to add a utility like `evidenceMatchers`: 
```go
	{
		Class:    "haproxy-binary",
		FileGlob: "**/haproxy",
		EvidenceMatcher: evidenceMatchers(
			fileContentsVersionMatchers(`(?m)HA-Proxy version (?P<version>[0-9]+\.[0-9]+\.[0-9]+)`),
			fileContentsVersionMatchers(`(?m)(?P<version>[0-9]+\.[0-9]+\.[0-9]+)-[0-9a-zA-Z]{7}.+HAProxy version`),
		),
		Package: "haproxy",
		PURL:    mustPURL("pkg:generic/haproxy@version"),
		CPEs:    singleCPE("cpe:2.3:a:haproxy:haproxy:*:*:*:*:*:*:*:*"),
	},
```

With:
```go
func evidenceMatchers(matchers ...evidenceMatcher) evidenceMatcher {
	return func(classifier classifier, reader source.LocationReadCloser) ([]pkg.Package, error) {
		for _, matcher := range matchers {
			pkgs, err := matcher(classifier, reader)
			if err != nil {
				return nil, err
			}
			if pkgs != nil {
				return pkgs, nil
			}
		}
		return nil, nil
	}
}
```

Either of these would get rid of the `-variant#` suffixes, which I think is preferable. Thoughts?

Reply:
much more elegant. I'll update the PR